### DISCUSSION TITLE
Ownership and Scope of Responsibility

### DISCUSSION THREAD

**Discussion by mokun** [mars-sim/mars-sim#315] [Ideas]
@bevans2000 @Urwumpe 

I have a few roadmap showcasing where mars-sim could be heading. 

The following is one of them.

Say we have 2 computers in a network running mars-sim as servers. 

We may also want to designate each machine to handle a certain geographical regions for simplicity.

So, computer 1 will handle the northern hemisphere, computer 2 will handle the southern hemisphere.

Each of them may also be responsible for the backend simulation of some of the global aspects of Mars such as weather, road constructions, mineral exploration or water discovery, satellites passing, resupply missions arrival, rocket launches, etc.

Player 1 is the CEO of corporation 1, he owns settlement A and remote station S. It runs on computer 3 as a client.

Player 2 is the CEO of corporation 2, he owns settlement B and remote station N. It runs on computer 4 as a client.

This introduces the concept of "jurisdiction". 

The simulation will have to keep track of the level or the kind of jurisdiction a player has 

Programmatically, we will somehow "tag"  settlement A and remote station S under player 1 and tag settlement B and remote station N under player 2.

Note that the remote station S could be in the south pole and the remote station N is in the north pole.

We would allow a player to remotely change things on a remote station.

To take one step further, we may even allow the concept of "private ownership".  

Player 1 may purchase or own certain private properties on a settlement B, even though settlement B is owned by player 2.

**Comment by Urwumpe** [comment]
Well, I see such complex things as ownership (I had nearly three weeks of lectures in civil law about just this topic) as a long-term goal, I think for the actual scenario, the term "Scope" or "responsibility" is more suitable.

So, for the beginning, it would be just enough to describe, which Commander is responsible for a settlement or other entity on the map (e.g. a rover travelling from settlement A to settlement B). Even transfering a rover to a new base would be fairly easy, since we could transfer such a responsibility easily in such a simple world.

The next, further step would already be more complex: Who owns something is not always the legal person who is in possession of it. We would need something like contracts to handle this - a remote base could be owned by player 1, but be the responsibility of player 2, for example. And from that point, it would be a small step to get from player 1 and player 2 actually being members of an organisation/guild/company that owns the remote base AND both having stakes into that organisation and decide about its strategy. 

I think, a good starting point for development there, would be making all events where responsibility is transfered explicit and visible. One kind of class should be responsible for all the logic behind responsibility transfers (Entity?) and nothing else should be able to bypass this.

Private property is only easy, as long as there is only one player. Or one player who is king (its good to be king). 

If you have the situation, that player 2 could gift something to player 1 while visiting his settlement, things already get complex. Or worse A large Settlement, where player 2 is subcommander of player 1.

**Reply by bevans2000** [reply]
I like the idea of "scope" of responsibilty that covers the Units. Starting with a new `Commander` class initially covering Settle ent would give us a start and lay down some principles. 
However the idea of multiple servers is a long way off I feel. We need to strengthen the concept of a 'Unit of Work' which updates a block of the simulation in isolation. This will take us towards updating the simulation in parallel within 1 server which is a stepping stone to parallel with multiple servers.
Nevertheless, one question I have is is there a need for multiple servers? Certainly there is a need for multiple clients but that is easier to develop. The current code base can be optimised to improve the speed and hence reduce the need for multiple servers. I ran my version of the code allowing remote access on a large AWS server today as a trial and it processed all Settlement in parallel pretty quickly. I will create a Pull Request tomorrow for this version.

**Reply by Urwumpe** [reply]
Well - I think yes, because of  the scope of the simulation. For < 1000 settlers, a small multicore simulation will be fine, but with ~10000 settlers, there should be a significant slowdown. Simulating a bigger colonization of Mars with 100,000 settlers in the end would be likely impossible to do with a useful msols/second ratio. 

Maybe we should add a heavy-weight simulation scenario just for finding out how terrible things really are. Getting rid of many global variables and functions should at least remove many bottlenecks and get a lot of speed-up.
